auth/src/main/java/com/firebase/ui/auth/ui/idp/WelcomeBackIdpPrompt.java

```diff
@@ -109,44 +112,69 @@ protected void onCreate(@Nullable Bundle savedInstanceState) {
             return;
         }
 
-        @StringRes int providerName;
+
+        String providerName;
+
+        String genericOAuthProviderId = config.getParams()
+                .getString(ExtraConstants.GENERIC_OAUTH_PROVIDER_ID);
+
         switch (providerId) {
             case GoogleAuthProvider.PROVIDER_ID:
                 GoogleSignInHandler google = supplier.get(GoogleSignInHandler.class);
                 google.init(new GoogleSignInHandler.Params(config, existingUser.getEmail()));
                 mProvider = google;
 
-                providerName = R.string.fui_idp_name_google;
+                providerName = getString(R.string.fui_idp_name_google);
                 break;
             case FacebookAuthProvider.PROVIDER_ID:
                 FacebookSignInHandler facebook = supplier.get(FacebookSignInHandler.class);
                 facebook.init(config);
                 mProvider = facebook;
 
-                providerName = R.string.fui_idp_name_facebook;
+                providerName = getString(R.string.fui_idp_name_facebook);
                 break;
             case TwitterAuthProvider.PROVIDER_ID:
                 TwitterSignInHandler twitter = supplier.get(TwitterSignInHandler.class);
                 twitter.init(null);
                 mProvider = twitter;
 
-                providerName = R.string.fui_idp_name_twitter;
+                providerName = getString(R.string.fui_idp_name_twitter);
                 break;
             case GithubAuthProvider.PROVIDER_ID:
                 ProviderSignInBase<AuthUI.IdpConfig> github =
                         supplier.get(GitHubSignInHandlerBridge.HANDLER_CLASS);
                 github.init(config);
                 mProvider = github;
 
-                providerName = R.string.fui_idp_name_github;
+                providerName = getString(R.string.fui_idp_name_github);
                 break;
             default:
-                throw new IllegalStateException("Invalid provider id: " + providerId);
+                if (TextUtils.equals(providerId, genericOAuthProviderId)) {
```

Comment:
If we're switching on `providerId` there's no need for this `equals` you can just do `case ExtraContstans.GENERIC_OAUTH_PROVIDER_ID` right?  Or am I missing something?

Reply:
Can't switch on GENERIC_OAUTH_PROVIDER_ID because the providerId is the actual value (e.g. "microsoft.com"). Unless I'm missing something? :smile:

Reply:
Ok I think many places I am being confused by the overload of the term "provider id" so I think we need to dedupe it.

For instance take Microsoft:
```java
 public static final class MicrosoftBuilder extends GenericOAuthProviderBuilder {
            private static final String PROVIDER_NAME = "Microsoft";

            public MicrosoftBuilder() {
                super(MICROSOFT_PROVIDER, PROVIDER_NAME, R.layout.fui_idp_button_microsoft);
            }
        }
```

Now looking at the super class of `MicrosoftBuilder`:

```java
 public static class GenericOAuthProviderBuilder extends Builder {

            GenericOAuthProviderBuilder(@NonNull String providerId,
                                        @NonNull String providerName,
                                        int buttonId) {
                super(GENERIC_OAUTH_PROVIDER);

                // ...

                if (!SUPPORTED_OAUTH_PROVIDERS.contains(providerId)) {
                    throw new IllegalStateException("You must specify a valid, " +
                            "supported provider ID.");
                }

                setProviderId(providerId);

                 // ...

            }
```


We passed `MICROSOFT_PROVIDER` for the `providerId` argument.  However we immediately call `super(GENERIC_OAUTH_PROVIDER)` which goes up to `IdpConfig.Builder` which has this constructor:

```java
           protected Builder(@SupportedProvider @NonNull String providerId) {
                if (!SUPPORTED_PROVIDERS.contains(providerId)) {
                    throw new IllegalArgumentException("Unknown provider: " + providerId);
                }
                mProviderId = providerId;
            }
```

So now we are saying the `providerId` is `GENERIC_OAUTH_PROVIDER`.  But then about 5 lines later we call `setProviderId` with a different value, which is the specific IDP constant which was previously called `MICROSOFT_PROVIDER`.

So ... which is it?  Why do we set the providerId to one thing and then immediately change it to another?  Maybe what we're trying to express here and throughout is that an `IdpConfig` has a "type" which controls the flow and a "providerId" which denotes a specific constant which can be compared to something the backend knows about?

Reply:
I used the GENERIC_OAUTH_PROVIDER as a placeholder ID in the supported providers set. The constructor of the builder checks that the provider belongs to this set and will throw an exception if it doesn't. It was just a way around this so we can support any provider. 

Since I already have a "SUPPORTED_OAUTH_PROVIDER" set, I'll just add that check and remove this logic. We'll have to re add it in the future if we decide to make it generic though. Hope that makes sense